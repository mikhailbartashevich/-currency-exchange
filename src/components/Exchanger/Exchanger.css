.Exchanger {
  font-family: 'Roboto';
  background: white;
  min-width: 250px;
  color: #7d7676;
  font-size: 12px;
}

.Exchanger_field {
  margin-bottom: 20px;
}

.Exchanger_label {
  margin-bottom: 5px;
  margin-right: 5px;
}

.Exchanger_rate {
  display: flex;
  justify-content: center;
}

.Exchanger_rate .Rate {
  margin-top: -30px;
}

.Exchanger_bottom-section,
.Exchanger_top-section {
  padding: 20px;
}

.Exchanger_bottom-section {
  background: #f5f5f5;
  min-height: 150px;
  position: relative;
}

.Exchanger_bottom-section .SwapIconButton {
  margin-top: -50px;
}

.Exchanger_submit {
  position: absolute;
  bottom: 20px;
  width: 250px;
}
