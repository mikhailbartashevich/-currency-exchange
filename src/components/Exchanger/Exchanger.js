import React from 'react';
import './Exchanger.css';
import { CurrencyInput } from '../CurrencyInput/CurrencyInput';
import { Rate } from '../Rate/Rate';
import { Balance } from '../Balance/Balance';
import { ExchangeFab } from '../ExchangeFab/ExchangeFab';
import { lifecycle, pure } from 'recompose';
import PropTypes from 'prop-types';

export const ExchangerRaw = props => (
  <div className="Exchanger">
    <div className="Exchanger_top-section">
      <CurrencyInput
        className="Exchanger_input"
        amount={props.inputAmount}
        currency={props.inputCurrency}
        currencyOptions={props.currencyOptions}
        actions={{
          changeAmount: props.changeInputAmount,
          changeCurrency: props.changeInputCurrency,
        }}
      />

      <div className="Exchanger_field">
        <span className="Exchanger_label">You have</span>
        <Balance
          amount={props.availableInputAmount}
          currency={props.inputCurrency}
        />
      </div>
    </div>
    <div className="Exchanger_bottom-section">
      <div className="Exchanger_rate">
        <Rate
          rate={props.currencyRate}
          inputCurrency={props.inputCurrency}
          outputCurrency={props.outputCurrency}
        />
      </div>

      <CurrencyInput
        amount={props.outputAmount}
        currency={props.outputCurrency}
        currencyOptions={props.currencyOptions}
        actions={{
          changeAmount: props.changeOutputAmount,
          changeCurrency: props.changeOutputCurrency,
        }}
      />

      <div className="Exchanger_field">
        <span className="Exchanger_label">You have</span>
        <Balance
          amount={props.availableOutputAmount}
          currency={props.outputCurrency}
        />
      </div>
      <div className="Exchanger_submit">
        <ExchangeFab
          variant="extended"
          aria-label="Exchange"
          onClick={props.exchange}
          disabled={
            props.availableInputAmount < props.inputAmount ||
            !Number(props.inputAmount)
          }
        >
          Exchange
        </ExchangeFab>
      </div>
    </div>
  </div>
);

ExchangerRaw.propTypes = {
  inputAmount: PropTypes.string,
  inputCurrency: PropTypes.shape({
    currency: PropTypes.string,
    symbol: PropTypes.string,
  }),
  availableInputAmount: PropTypes.number,
  changeInputAmount: PropTypes.func,
  changeInputCurrency: PropTypes.func,

  outputAmount: PropTypes.string,
  outputCurrency: PropTypes.shape({
    currency: PropTypes.string,
    symbol: PropTypes.string,
  }),
  availableOutputAmount: PropTypes.number,
  changeOutputAmount: PropTypes.func,
  changeOutputCurrency: PropTypes.func,

  exchange: PropTypes.func,
  currencyOptions: PropTypes.arrayOf(
    PropTypes.shape({
      currency: PropTypes.string,
      symbol: PropTypes.string,
    }),
  ),
};

export const Exchanger = pure(
  lifecycle({
    componentDidMount() {
      this.props.fetchRates &&
        this.props.fetchRates(
          this.props.inputCurrency,
          this.props.outputCurrency,
        );
      clearInterval(this.fetchInterval);
      this.fetchInterval = setInterval(
        () =>
          this.props.fetchRates(
            this.props.inputCurrency,
            this.props.outputCurrency,
          ),
        10000,
      );
    },

    componentWillUnmount() {
      clearInterval(this.fetchInterval);
    },
  })(ExchangerRaw),
);
