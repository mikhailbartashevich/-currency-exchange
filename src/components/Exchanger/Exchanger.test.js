import React from 'react';
import { shallow } from 'enzyme';
import { ExchangerRaw } from './Exchanger';

describe('Exchanger', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<ExchangerRaw debug />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with no props', () => {
    const component = shallow(<ExchangerRaw />);
    expect(component).toMatchSnapshot();
  });

  it('should render form with given props', () => {
    const inputCurrency = { currency: 'USD', symbol: '$' };
    const outputCurrency = { currency: 'EUR', symbol: '€' };
    const currencyOptions = [
      { currency: 'USD', symbol: '$' },
      { currency: 'EUR', symbol: '€' },
      { currency: 'GBP', symbol: '£' },
    ];
    const fetchRatesFn = jest.fn();
    const component = shallow(
      <ExchangerRaw
        inputAmount="10"
        inputCurrency={inputCurrency}
        outputAmount="10"
        outputCurrency={outputCurrency}
        currencyRate="0.6"
        currencyOptions={currencyOptions}
        availableInputAmount={20}
        availableOutputAmount={30}
        fetchRates={fetchRatesFn}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('button click should send exchange event', () => {
    const clickFn = jest.fn();
    const inputCurrency = { currency: 'USD', symbol: '$' };
    const outputCurrency = { currency: 'EUR', symbol: '€' };
    const component = shallow(
      <ExchangerRaw
        inputAmount="10"
        inputCurrency={inputCurrency}
        outputAmount="10"
        outputCurrency={outputCurrency}
        currencyRate="0.6"
        availableInputAmount={20}
        availableOutputAmount={30}
        exchange={clickFn}
      />,
    );
    component.find('pure(WithStyles(ExchangeFabRaw))').simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});
