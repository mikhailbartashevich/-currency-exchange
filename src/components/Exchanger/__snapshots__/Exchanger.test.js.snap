// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchanger should render correctly in "debug" mode 1`] = `
<div
  className="Exchanger"
>
  <div
    className="Exchanger_top-section"
  >
    <pure(CurrencyInputRaw)
      actions={
        Object {
          "changeAmount": undefined,
          "changeCurrency": undefined,
        }
      }
      className="Exchanger_input"
    />
    <div
      className="Exchanger_field"
    >
      <span
        className="Exchanger_label"
      >
        You have
      </span>
      <pure(BalanceRaw) />
    </div>
  </div>
  <div
    className="Exchanger_bottom-section"
  >
    <div
      className="Exchanger_rate"
    >
      <pure(RateRaw) />
    </div>
    <pure(CurrencyInputRaw)
      actions={
        Object {
          "changeAmount": undefined,
          "changeCurrency": undefined,
        }
      }
    />
    <div
      className="Exchanger_field"
    >
      <span
        className="Exchanger_label"
      >
        You have
      </span>
      <pure(BalanceRaw) />
    </div>
    <div
      className="Exchanger_submit"
    >
      <pure(WithStyles(ExchangeFabRaw))
        aria-label="Exchange"
        disabled={true}
        variant="extended"
      >
        Exchange
      </pure(WithStyles(ExchangeFabRaw))>
    </div>
  </div>
</div>
`;

exports[`Exchanger should render correctly with no props 1`] = `
<div
  className="Exchanger"
>
  <div
    className="Exchanger_top-section"
  >
    <pure(CurrencyInputRaw)
      actions={
        Object {
          "changeAmount": undefined,
          "changeCurrency": undefined,
        }
      }
      className="Exchanger_input"
    />
    <div
      className="Exchanger_field"
    >
      <span
        className="Exchanger_label"
      >
        You have
      </span>
      <pure(BalanceRaw) />
    </div>
  </div>
  <div
    className="Exchanger_bottom-section"
  >
    <div
      className="Exchanger_rate"
    >
      <pure(RateRaw) />
    </div>
    <pure(CurrencyInputRaw)
      actions={
        Object {
          "changeAmount": undefined,
          "changeCurrency": undefined,
        }
      }
    />
    <div
      className="Exchanger_field"
    >
      <span
        className="Exchanger_label"
      >
        You have
      </span>
      <pure(BalanceRaw) />
    </div>
    <div
      className="Exchanger_submit"
    >
      <pure(WithStyles(ExchangeFabRaw))
        aria-label="Exchange"
        disabled={true}
        variant="extended"
      >
        Exchange
      </pure(WithStyles(ExchangeFabRaw))>
    </div>
  </div>
</div>
`;

exports[`Exchanger should render form with given props 1`] = `
<div
  className="Exchanger"
>
  <div
    className="Exchanger_top-section"
  >
    <pure(CurrencyInputRaw)
      actions={
        Object {
          "changeAmount": undefined,
          "changeCurrency": undefined,
        }
      }
      amount="10"
      className="Exchanger_input"
      currency={
        Object {
          "currency": "USD",
          "symbol": "$",
        }
      }
      currencyOptions={
        Array [
          Object {
            "currency": "USD",
            "symbol": "$",
          },
          Object {
            "currency": "EUR",
            "symbol": "€",
          },
          Object {
            "currency": "GBP",
            "symbol": "£",
          },
        ]
      }
    />
    <div
      className="Exchanger_field"
    >
      <span
        className="Exchanger_label"
      >
        You have
      </span>
      <pure(BalanceRaw)
        amount={20}
        currency={
          Object {
            "currency": "USD",
            "symbol": "$",
          }
        }
      />
    </div>
  </div>
  <div
    className="Exchanger_bottom-section"
  >
    <div
      className="Exchanger_rate"
    >
      <pure(RateRaw)
        inputCurrency={
          Object {
            "currency": "USD",
            "symbol": "$",
          }
        }
        outputCurrency={
          Object {
            "currency": "EUR",
            "symbol": "€",
          }
        }
        rate="0.6"
      />
    </div>
    <pure(CurrencyInputRaw)
      actions={
        Object {
          "changeAmount": undefined,
          "changeCurrency": undefined,
        }
      }
      amount="10"
      currency={
        Object {
          "currency": "EUR",
          "symbol": "€",
        }
      }
      currencyOptions={
        Array [
          Object {
            "currency": "USD",
            "symbol": "$",
          },
          Object {
            "currency": "EUR",
            "symbol": "€",
          },
          Object {
            "currency": "GBP",
            "symbol": "£",
          },
        ]
      }
    />
    <div
      className="Exchanger_field"
    >
      <span
        className="Exchanger_label"
      >
        You have
      </span>
      <pure(BalanceRaw)
        amount={30}
        currency={
          Object {
            "currency": "EUR",
            "symbol": "€",
          }
        }
      />
    </div>
    <div
      className="Exchanger_submit"
    >
      <pure(WithStyles(ExchangeFabRaw))
        aria-label="Exchange"
        disabled={false}
        variant="extended"
      >
        Exchange
      </pure(WithStyles(ExchangeFabRaw))>
    </div>
  </div>
</div>
`;
